Add `authctl user set-shell` command

adombeck opened the pull request:
Allow changing the shell of authd users.

Closes #939
UDENG-7089

adombeck committed:
Add gRPC method SetShell

adombeck committed:
Add `authctl user set-shell` command

3v1n0 commented on internal/users/manager.go:
```diff
@@ -664,6 +664,33 @@ func checkHomeDirOwner(home string, uid, gid uint32) error {
 }
 
+// SetShell sets the shell for the given user.
+func (m *Manager) SetShell(username, shell string) (warnings []string, err error) {
```
Shouldn't we also check if the shell is allowed in `/etc/shells`?

adombeck replied:
> we do have /etc

In Ubuntu, yes, but we can also look at it from an authd upstream perspective. Other distros which don't have `/etc/` might also want to use authd.

>  and we do not care about system users

Same as systemd-homed AFAIU.

3v1n0 replied:
Don't know, not fully convinced... Maybe we could ignore the check if /etc isn't there, but otherwise I would like to follow the distro preferences in this regards.

adombeck committed:
Test SetShell methods in db and users packages

adombeck committed:
docs: Generate docs for `authctl user set-shell`

adombeck committed:
Rename TestIsRequestFromRoot -> TestCheckRequestIsFromRoot

The function that's tested is CheckRequestIsFromRoot.

adombeck committed:
Rename peerCredsInfo to peerAuthInfo

... because the field in peer.Peer is named AuthInfo.

Also improves the comment of WithUnixPeerCreds.

adombeck committed:
Test the SetShell method of the user service

adombeck committed:
Improve error message

The grpc.NewClient does not actually try to connect to authd, it just
sets up the client but doesn't do any I/O.

adombeck committed:
SetShell: Only allow root to set a user's shell

The original plan was to allow users to change their own shell without
root permissions, like chsh does. However, chsh requires the user to
authenticate before it allows to change the shell.

We should also be able to do that by letting the user authenticate to
the broker. That's a lot more effort than originally planned though,
so for now we only allow root to change user shells.

adombeck committed:
SetShell: Allow root to set an invalid shell

But display a warning, similar to the behavior of chsh.

adombeck committed:
SetUserID: Improve warning messages

adombeck committed:
TestSetUserID: Fix entire warning message replaced

The warning message in the golden file still didn't have the "Warning:"
prefix we now add to most of our warnings. That's because the tests
replaced the entire warning message.

Fix by only replacing the parts which need to be replaced.

3v1n0 commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+	if err != nil {
+		return err
+	}
+	for _, allowedShell := range strings.Split(string(shells), "\n") {
```
What about using `bufio.NewScanner()`

adombeck replied:
done

3v1n0 commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+			// Skip comments
+			continue
+		}
+		if allowedShell == shell {
```
This implies the full shell path, that is good, but do we do any transformation client-side? IIRC `chsh` also requires a full path, but just to be sure.

adombeck replied:
> but do we do any transformation client-side?

do you mean in authctl? then no, it passes the shell argument as-is to the server: https://github.com/canonical/authd/blob/ee009434f9d9eaea27f258d61aafb9bc8aaf90f3/cmd/authctl/user/set-shell.go#L21-L36

> IIRC chsh also requires a full path

correct:

```
chsh $USER --shell bash
Password: 
chsh: bash is an invalid shell
```

3v1n0 commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	for _, r := range value {
+		if r < 32 || r == 127 {
```
Isn't there any function allowing to check the same?

adombeck replied:
there is. I'm now using `unicode.IsControl`.

3v1n0 commented on internal/envutils/envutils.go:
```diff
@@ -0,0 +1,46 @@
+	if strings.ContainsRune(value, '\x00') {
+		return nil, fmt.Errorf("invalid value: %q", value)
+	}
+
```
Maybe also check for valid utf value?

adombeck replied:
I'd prefer to keep the same behavior as `os.Setenv` which also only checks for `0`.

adombeck commented on internal/services/permissions/permissions.go:
```diff
@@ -40,18 +40,50 @@ func New(args ...Option) Manager {
+
+// CheckRequestIsFromRootOrUID checks if the current gRPC request is from a root
+// user or a specified user and returns an error if not.
+func (m Manager) CheckRequestIsFromRootOrUID(ctx context.Context, uid uint32) (err error) {
```
I dropped it

adombeck commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+
+	// Check if the shell is in the list of allowed shells in /etc/shells
+	shells, err := os.ReadFile("/etc/shells")
+	if err != nil {
```
I would keep it as is until it's causing an issue for anyone

adombeck commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+		return err
+	}
+	for _, allowedShell := range strings.Split(string(shells), "\n") {
+		if allowedShell[0] == '#' {
```
outdated

adombeck commented on internal/users/manager.go:
```diff
@@ -665,6 +665,33 @@ func checkHomeDirOwner(home string, uid, gid uint32) error {
+		return fmt.Errorf("shell %q is not executable", shell)
+	}
+
+	if err = m.db.SetShell(username, shell); err != nil {
```
I think `lckpwdf()` is only necessary when setting the UID or name. I would prefer to avoid using it unnecessarily.

3v1n0 replied:
Yeah, well technically it's a lock to update anything, but in this case could just avoid a shell not being loaded instead of another. Indeed not a big deal.

adombeck commented on internal/envutils/envutils.go:
```diff
@@ -0,0 +1,46 @@
+// Package envutils provides utilities for manipulating string slices representing environment variables.
+package envutils
```
That's a nice optimization but I don't think it's worth doing now because currently we don't call the `Getenv` and `Setenv` methods often for the same environment. I'll add a comment though.

adombeck commented on cmd/authctl/group/group_test.go:
```diff
@@ -32,6 +32,7 @@ func TestGroupCommand(t *testing.T) {
 
 			//nolint:gosec // G204 it's safe to use exec.Command with a variable here
 			cmd := exec.Command(authctlPath, append([]string{"group"}, tc.args...)...)
+			cmd.Env = []string{testutils.CoverDirEnv()}
```
IIUC, that comment was about the `-cover` flag passed to `go build`. Here, we pass the `GOCOVERDIR` environment variable when running authctl. However, it looks like we didn't address that comment and we currently don't pass `-cover` to `go build`. I wonder if that's really since we still get coverage data 🤔

adombeck commented on internal/envutils/envutils.go:
```diff
@@ -0,0 +1,46 @@
+	for _, kv := range env {
+		if strings.HasPrefix(kv, key+"=") {
+			return strings.TrimPrefix(kv, key+"=")
+		}
```
done

adombeck commented on internal/users/userutils.go:
```diff
@@ -0,0 +1,68 @@
+		return errors.New("value must be valid UTF-8")
+	}
+
+	if strings.ContainsRune(value, ':') {
```
I think `:` is generally not allowed in any of the fields of a user record because it would break tools which use the output of `getent passwd` and split it by `:`

adombeck commented on internal/users/manager.go:
```diff
@@ -665,6 +665,33 @@ func checkHomeDirOwner(home string, uid, gid uint32) error {
+	stat, err := os.Stat(shell)
+	if errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("shell %q does not exist", shell)
+	}
```
done

adombeck commented on internal/proto/authd/authd.proto:
```diff
@@ -202,6 +202,10 @@ message SetShellRequest {
 }
 
+message SetShellResponse {
+  repeated string warnings = 1;
```
Out of scope of this PR, something to consider when we add i18n support

3v1n0 replied:
Well, knida, we could just expose an error message ID and support it on client side?

Or would you prefer to handle this all together?

adombeck replied:
yeah I think we should add the needed functionality in all components once we tackle i18n support

adombeck committed:
man/authctl: Re-generate to add set-shell

3v1n0 commented on internal/proto/authd/authd.pb.go.orig:
This should not be needed here no?

adombeck replied:
oops

adombeck replied:
fixed

3v1n0 commented on internal/proto/authd/authd.proto.orig:
Also drop this.

adombeck replied:
yeah I'm on it

adombeck replied:
done

3v1n0 commented on internal/proto/authd/authd_grpc.pb.go.orig:
And this

adombeck replied:
done
